### ISSUE TITLE
[IPTV Simple Client] EPG loads from server on every Kodi restart

### ISSUE THREAD

**Issue by Psycho-A** [xbmc/xbmc#19107]
I know this problem has been discussed a lot around, but I haven't found a coherent answer.

I have configured the IPTV Client to retrieve the m3u playlist and xmtv guide from external servers. I also set the TV guide update interval to 600 hours. But why Kodi does loading EPG every time I launch it? I thought the update interval is enough to say how program should work with recieving new data. But it's not! I have m3u playlist with 200 positions and large .xml.gz tv-guide which is about 100 mb unpacked, end every Kodi startup I see “Importing Guide from clients” during about 5 minutes, and program even may totally hang. Also, judging by the constant activity disk indicator during this time, it is persistently writing something to my SD card. Guys, I run Kodi on my Raspberry Pi 3 with a media that have a limited write cycle, and this is very critical to it!

I don't want constant updates - I only want the comfortable startup and do updates at the interval I specified in settings, because my TV guide is only updated once a week. Please make it possible to users to completely disable this "feature" for those who don't use it. I only use the IPTV Client to watch our local provider channels, and they have no PVR features or something else. Judging from the discussions, the problem is already more than 6 years old...

**Comment by phunkyfish** [comment]
Kodi 19 was changed recently so that it would only load when needed. However another change was made recently as kodi's copy of the data and the iptvsimple's are separate. I.e. the add-on has no permanent store of data and require's it to support catchup. 

But it should cache your EPG data if it's external. So once it doesn't change it shouldn't download.

**Comment by basilgello** [comment]
@phunkyfish I suggest implementing HTTP `If-Modified-Since` header support (it's trivial in cURL) and treating `304 Not modified` on both M3U and XMLTV requests (if configured, ofc!) then report to Kodi that update can be skipped.

**Comment by phunkyfish** [comment]
I believe under the hood this is what stat ok the file does in XBMC’s XFile.

**Comment by phunkyfish** [comment]
I imagine the issue is that the file has changed on each kodi start and therefore it gets downloaded.

**Comment by wsnipex** [comment]
if the server set correct Cache Control headers, a re-download could be prevented, even if the file changed.
Note that I didn't check if we currently handle that correctly, but if not, it should be fixed.

**Comment by basilgello** [comment]
> if the server set correct Cache Control headers, a re-download could be prevented, even if the file changed.

Apart from 'Cache-Control', there are also 'Expired' and 'If-Modified-Since' headers, and I think it's worth handling these too. They are part of IETF W3C standard and are supported by all major browsers and servers.

I suggest the following usecase:

* If we have a modification time of file resource, we just send 'If-Modified-Since: Sat 31 Dec 2020 00:00:00 UTC' (mtime formatted like that) and if the HTTP response code is '304 Not Modified' then we assume the remote resource is intact compared to our local copy.

'Cache-Control' and 'Expired' are less trivial to check because we need to store the values of that headers somewhere. This is not OK for Kodi-wise general code like CCURLFile but is OK for PVR addons like IPTVSimple. There is literally nothing preventing the PVR client from storing the contents of 'Cache-Control: max-age=XXXXXX' (seconds) or 'Expired: XXXXXXX' (timestamp) and calculate the projected expiration timestamp to base decision on.

**Comment by Psycho-A** [comment]
I don't know, were files on server changed, but it doesn't look like that because Kodi redownloads it everytime even if launching interval is just 1 minute, and this happens with any XMLTV service link I provided. Of course caching on disk is enabled both for m3u and xmltv. I don't know about the things described above, but I think that simplest way is that user's "update interval" setting would have the highest priority over other behavior modes, even if there are file updates on the server. Just don't tell the program to do anything (if files already presented on disk/database; if not, download them once) until the specified time period has elapsed. On my feeling this setting is just malfunctioned right now.

**Comment by ksooo** [comment]
> I think that simplest way is that user's "update interval" setting would have the highest priority over other behavior modes

 Have you also tested with v19 RC1? Maybe we are talking about things that got fixed meanwhile?

**Comment by phunkyfish** [comment]
To be clear. Iptvsimple needs to download the epg each time it starts. It is a requirement for catchup. What I can do is make this conditional on catchup being enabled.

**Comment by basilgello** [comment]
@ksooo yes the problem is still there on tonight trunk. But again it is iptvsimple issue, and even of catchup is enabled, it is still worth checking if remote IPTV playlists and EPG were updated.

**Comment by phunkyfish** [comment]
@Psycho-A please try a testbuild from here: https://jenkins.kodi.tv/blue/organizations/jenkins/kodi-pvr%2Fpvr.iptvsimple/detail/PR-469/2/artifacts

If running RC1 it should only download the xmltv if kodi requests it instead of every startup. Note that it will only work like this if the catchup feature is disabled.

**Comment by Psycho-A** [comment]
@phunkyfish 
Thanks, but I can't use v19 RC1 for now since I'm on closed RPi OS with Kodi integrated (it can't update itself), so I need to wait the OS update. Can your updated iptvsimple plugin will work correctly with v18.7? And what is "cactchup feature"?

**Comment by phunkyfish** [comment]
Ah, ok. Then I think you must wait. There are no further updates to Kodi 18 and you can’t use the Kodi 19 add-on with Kodi 18.

Catchup allows you to go back in time (if your providers supports it). So you can timeshift to before the time you started playing. There is also regular timeshift in kodi 19’s iptvsimple that allows you go back till any time since you started playing too.

**Comment by basilgello** [comment]
Or upgrade to Raspbian if it Raspberry Pi. Bullseye (testing) has 19.0 RC in main repository.

**Comment by basilgello** [comment]
I initially suspected that the `Last-Modified` header gets ignored by `kodi::vfs::stat_file()` but I was wrong:

```shell
# stdbuf tshark -i lo -V -Y "http.request || http.response" -O http port 8081 | grep "Expert Info"
Running as user "root" and group "root". This could be dangerous.
Capturing on 'Loopback: lo'
        [Expert Info (Chat/Sequence): HEAD /divan/playlist HTTP/1.1\r\n]
        [Expert Info (Chat/Sequence): HTTP/1.1 200 OK\r\n]
        [Expert Info (Chat/Sequence): HEAD /divan/epg HTTP/1.1\r\n]
        [Expert Info (Chat/Sequence): HTTP/1.1 200 OK\r\n]
```

The playlist is still loaded from cache here: https://github.com/kodi-pvr/pvr.iptvsimple/blob/Matrix/src/iptvsimple/utilities/FileUtils.cpp#L173 so we need to tell Kodi PVR engine that the channel and EPG data to be supplied by the particular PVR client is not modified.

@phunkyfish @ksooo I recall you did some kind of flagging the need for EPG updates?

**Comment by phunkyfish** [comment]
I don’t think this is anything to do with the kodi PVR engine. It will call for EPG data when it needs it only.

The playlist will always be loaded on start. The EPG data will be loaded when kodi asks for or if catchup is enabled.

**Comment by alpha7504** [comment]
This issue is still present in Kodi 19.3 matrix... EPG loads every time kodi is restarted, and it slows down the overall performance... I have set it to update every 10 hours yet this setting is being ignored.

**Comment by tbrek** [comment]
> This issue is still present in Kodi 19.3 matrix... EPG loads every time kodi is restarted, and it slows down the overall performance... I have set it to update every 10 hours yet this setting is being ignored.

Setting it up to fetch every 10h is diff than loading on startup which. Yours now loads on startup and every 10h. 

To overcome the issue I'm caching mine on R-Pi which is beinbf update by cronjob every 24h. It is much faster to fetch from local source then when Kodi starts.

**Comment by phunkyfish** [comment]
> This issue is still present in Kodi 19.3 matrix... EPG loads every time kodi is restarted, and it slows down the overall performance... I have set it to update every 10 hours yet this setting is being ignored.

Do you use catchup/archive? If you did it must be loaded on startup as the add-on requires that state when it starts up. If you don’t it should not try to load on a restart unless the EPG update window has passed.

Note that kodi requesting an EPG update and the addon updating are not quite the same thing.

**Comment by alpha7504** [comment]
> Setting it up to fetch every 10h is diff than loading on startup which. Yours now loads on startup and every 10h.
Thanks for your reply..
I am using it on android device. Even if i set to to update every 24hours, EPG still updates on kodi restart.

> Do you use catchup/archive? If you did it must be loaded on startup as the add-on requires that state when it starts up. If you don’t it should not try to load on a restart unless the EPG update window has passed.
Thanks for your reply..

I don't use catchup... it is turned off along with time shift. I have tried about everything i could think of... but still EPG loads from server on every kodi restart. In the end i shortened the playlist to very minimum channels so it doesn't hogs the system much even if EPG reloads every time..

**Comment by phunkyfish** [comment]
Does kodi run on the device constantly? Or is it started only when needed?

**Comment by alpha7504** [comment]
It starts when only needed... I set 24 hour (1440 min) in pvr and live tv settings>>guide.

**Comment by phunkyfish** [comment]
> It starts when only needed... I set 24 hours (1440 mins) in pvr and live tv settings>>guide.

Ok, and does more than 24 hours pass between the times you start kodi?

Or said another way if you shutdown kodi and then start it immediately does it still try to load the EPG?

**Comment by alpha7504** [comment]
> Or said another way if you shutdown kodi and then start it immediately does it still try to load the EPG?
Yes... It does. Every time I shutdown Kodi and start it again immediately, EPG loads again.
